lightning/src/ln/payment_tests.rs

```diff
@@ -1304,6 +1305,107 @@ fn onchain_failed_probe_yields_event() {
 	assert!(!nodes[0].node.has_pending_payments());
 }
 
+#[test]
+fn preflight_probes_yield_event_and_skip() {
+	let chanmon_cfgs = create_chanmon_cfgs(5);
+	let node_cfgs = create_node_cfgs(5, &chanmon_cfgs);
+
+	// We alleviate the HTLC max limit, as otherwise we'd always be limited through that.
+	let mut no_htlc_limit_config = test_default_channel_config();
+	no_htlc_limit_config.channel_handshake_config.max_inbound_htlc_value_in_flight_percent_of_channel = 100;
+
+	let user_configs = std::iter::repeat(no_htlc_limit_config).take(5).map(|c| Some(c)).collect::<Vec<Option<UserConfig>>>();
+	let node_chanmgrs = create_node_chanmgrs(5, &node_cfgs, &user_configs);
+	let nodes = create_network(5, &node_cfgs, &node_chanmgrs);
+
+	// Setup channel topology:
+	//                    (30k:0)- N2 -(1M:0)
+	//                   /                  \
+	//  N0 -(100k:0)-> N1                    N4
+	//                   \                  /
+	//                    (70k:0)- N3 -(1M:0)
+	//
+	let first_chan_update = create_announced_chan_between_nodes_with_value(&nodes, 0, 1, 100_000, 0).0;
+	create_announced_chan_between_nodes_with_value(&nodes, 1, 2, 30_000, 0);
+	create_announced_chan_between_nodes_with_value(&nodes, 1, 3, 70_000, 0);
+	create_announced_chan_between_nodes_with_value(&nodes, 2, 4, 1_000_000, 0);
+	create_announced_chan_between_nodes_with_value(&nodes, 3, 4, 1_000_000, 0);
+
+	let mut invoice_features = Bolt11InvoiceFeatures::empty();
+	invoice_features.set_basic_mpp_optional();
+
+	let mut payment_params = PaymentParameters::from_node_id(nodes[4].node.get_our_node_id(), TEST_FINAL_CLTV);
+	if let Payee::Clear { features, .. } = &mut payment_params.payee {
+		*features = Some(invoice_features);
+	}
+
+	let route_params = RouteParameters { payment_params, final_value_msat: 80_000_000 };
+	let res = nodes[0].node.send_preflight_probes(route_params).unwrap();
+
+	// We check that only one probe was sent, the other one was skipped due to limited liquidity.
+	assert_eq!(res.len(), 1);
+	let log_msg = format!("Skipped sending payment probe to avoid putting channel {} under the liquidity limit.",
+		first_chan_update.contents.short_channel_id);
+	node_cfgs[0].logger.assert_log_contains("lightning::ln::channelmanager", &log_msg, 1);
```

Comment:
It's a bit fragile and dependent on the implementation rather than the interface. Would prefer to have one test setup that doesn't skip any paths and another test with a different setup that does skip a path. I guess that may be a bit verbose without DRYing up the message handling.